src/main/java/org/fenixedu/academic/domain/student/StatuteType.java

```diff
@@ -0,0 +1,200 @@
+        if (isCodeUsed(code)) {
+            throw new DomainException("error.StatuteType.code.alreadyUsed");
+        }
+        setCode(code);
+        setName(name);
+        setGrantsWorkingStudentStatute(grantsWorkingStudentStatute);
+        setAssociativeLeaderStatute(associativeLeaderStatute);
+        setSpecialSeasonGrantedByRequest(specialSeasonGrantedByRequest);
+        setGrantOwnerStatute(grantOwnerStatute);
+        setSeniorStatute(seniorStatute);
+        setHandicappedStatute(handicappedStatute);
+        setActive(active);
+
+        setExplicitCreation(explicitCreation);
+        setVisible(visible);
+        setSpecialSeasonGranted(specialSeasonGranted);
+
+        checkRules();
+    }
+
+    @Override
+    public void setCode(String code) {
+        if (StringUtils.isEmpty(code)) {
+            return;
+        }
+
+        super.setCode("");
+        if (isCodeUsed(code)) {
+            throw new DomainException("error.StatuteType.code.alreadyUsed");
+        }
+        super.setCode(code);
+    }
+
+    private boolean isCodeUsed(String code) {
+        for (StatuteType statuteType : Bennu.getInstance().getStatuteTypesSet()) {
+            if (code.equals(statuteType.getCode())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    protected void checkRules() {
+        check(getName(), "error.StatuteType.name.required");
+    }
+
+    private static void check(final Object obj, final String message, final String... args) {
+        if (obj == null) {
+            throw new DomainException(message, args);
+        }
+    }
+
+    public boolean isGrantsWorkingStudentStatute() {
+        return getGrantsWorkingStudentStatute() != null && getGrantsWorkingStudentStatute();
+    }
+
+    @Deprecated
+    public boolean isAssociativeLeaderStatute() {
+        return getAssociativeLeaderStatute() != null && getAssociativeLeaderStatute();
+    }
+
+    @Deprecated
+    public boolean isSpecialSeasonGrantedByRequest() {
+        return getSpecialSeasonGrantedByRequest() != null && getSpecialSeasonGrantedByRequest();
+    }
+
+    public boolean isGrantOwnerStatute() {
+        return getGrantOwnerStatute() != null && getGrantOwnerStatute();
+    }
+
+    public boolean isSeniorStatute() {
+        return getSeniorStatute() != null && getSeniorStatute();
+    }
+
+    public boolean isHandicappedStatute() {
+        return getHandicappedStatute() != null && getHandicappedStatute();
+    }
+
+    @Deprecated
+    public boolean isActive() {
+        return getActive() != null && getActive();
+    }
+
+    public boolean isDeletable() {
+        return getStudentStatutesSet().isEmpty();
+    }
+
+    @Atomic
+    public void delete() {
+        if (!isDeletable()) {
+            throw new DomainException("error.StatuteType.deletion.not.possible");
+        }
+
+        setRootDomainObject(null);
+        deleteDomainObject();
+    }
+
+    public static Collection<StatuteType> readAll() {
```

Comment:
This should just return a `Stream`, as it is mostly used to filter and lookup values.